starstream_sys/src/lib.rs

```diff
@@ -90,13 +90,23 @@ unsafe extern "C" {
 
     #[link_name = "starstream_this_code"]
     pub safe fn this_code() -> CodeHash;
+
+    #[link_name = "starstream_keccak256"]
+    fn precompile_keccak256(buf: *const u8, len: usize, result: *mut u8);
 }
 
 #[inline]
 pub fn log(buf: &[u8]) {
     unsafe { starstream_log(buf.as_ptr(), buf.len()) }
 }
 
+#[inline]
+pub fn keccak256(buf: &[u8]) -> [u8; 32] {
+    let mut out = [0u8; 32];
+    unsafe { precompile_keccak256(buf.as_ptr(), buf.len(), out.as_mut_slice().as_mut_ptr()) };
+    out
+}
+
```

Comment:
does adding precompiles in this way properly compile to WASM in a way that it can be included in the ZK proofs once we connect them?

generally, in Starstream, I think we'll want to have some kind of "library" script type which is a program that is not a UTXO (stateless), but also not a coordination script (not the entry point of a transaction). It would be useful to have so that we can have library scripts like keccak256, but also things like VDF calculation

Reply:
I may be wrong here, but as far as I understand this will run in the host (wasmi now), which means it runs in Rust (like a "system call").

But right now if you wanted to run the function in wasm you can just include the dependency in your contract and call the function, and just compile everything together to wasm, embedding the keccak implementation.

That said, for something like this, doesn't it actually make more sense to generate a custom circuit optimized to keccak, instead of actually proving the wasm execution with the overhead of the vm? At least that's what I've seen in other zkvms.

Reply:
Right now, as Enzo says, this runs in the host like a system call. I'm still hammering on the effect scheduler but it should ideally make it easy for us to change this to either be a WASM module for ZK or a true precompile (custom circuit) for ZK if/when we need to.

> "env" is a weird name for the shared module.

"env" is the default module name that Rust/Emscripten/etc use for the C namespace if not overridden. We'll probably end up tweaking it later, but this keeps things simple for now.

Reply:
I guess if we really want to run it in wasm we could also call another wasmi interpreter inside the func_wrap handler too. That way it wouldn't have to be statically linked.